source/ArrowKeyStepper/ArrowKeyStepper.js

```diff
@@ -40,45 +40,35 @@ export default class ArrowKeyStepper extends React.PureComponent<
     scrollToRow: 0,
   };
 
+  state = {
+    scrollToColumn: 0,
+    scrollToRow: 0,
+  };
+
   _columnStartIndex = 0;
   _columnStopIndex = 0;
   _rowStartIndex = 0;
   _rowStopIndex = 0;
 
-  constructor(props: Props) {
-    super(props);
-
-    this.state = {
-      scrollToColumn: props.scrollToColumn,
-      scrollToRow: props.scrollToRow,
-    };
-  }
-
-  componentWillReceiveProps(nextProps: Props) {
-    if (this.props.isControlled) {
-      return;
+  static getDerivedStateFromProps(
+    nextProps: Props,
+    prevState: State,
+  ): $Shape<State> {
+    if (nextProps.isControlled) {
+      return null;
     }
 
-    const {scrollToColumn, scrollToRow} = nextProps;
-
-    const {
-      scrollToColumn: prevScrollToColumn,
-      scrollToRow: prevScrollToRow,
-    } = this.props;
-
     if (
-      prevScrollToColumn !== scrollToColumn &&
-      prevScrollToRow !== scrollToRow
+      nextProps.scrollToColumn !== prevState.scrollToColumn ||
+      nextProps.scrollToRow !== prevState.scrollToRow
     ) {
-      this.setState({
-        scrollToColumn,
-        scrollToRow,
-      });
-    } else if (prevScrollToColumn !== scrollToColumn) {
-      this.setState({scrollToColumn});
-    } else if (prevScrollToRow !== scrollToRow) {
-      this.setState({scrollToRow});
+      return {
+        scrollToColumn: nextProps.scrollToColumn,
+        scrollToRow: nextProps.scrollToRow,
+      };
```

Comment:
This is more a comment on the new React API than anything, but I have to admit I found this hard to read, especially because `_setScrollPosition()` ends up calling `this.setState()`, `_updateScrollPositionForScrollToCell()` ends up maybe calling `_setScrollPosition()`, and this lifecycle ends up returning an object that is "magically" set to state. All 3 methods handle somewhat overlapping responsibilities yet they all do so in a different way.

(The added ending `return null` in `_setScrollPosition()` also threw me for a second, because I associated returning with `calculateDerivedStateFromProps`. So that method either ends with `setState` (which returns nothing) or returning nothing.)

I'm not sure what the best way to make this clearer would be (or if more experience would be enough). Maybe type signatures? I don't really use flow -- does it have a way to signify "this method will have a side effect via setting state" or something?

Reply:
I think this comment is more pertaining to `CollectionView` than `ArrowKeyStepper`?

I think the null return from `_setScrollPosition` was just an oversight on my part. 😄

Reply:
Ah, yeah -- comment should be there instead :)